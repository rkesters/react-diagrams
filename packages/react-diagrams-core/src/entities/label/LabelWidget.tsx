import * as React from 'react';
import { DiagramEngine } from '../../DiagramEngine';
import { LabelModel } from './LabelModel';
import styled from '@emotion/styled';
import { get } from 'lodash';

export interface LabelWidgetProps {
	engine: DiagramEngine;
	label: LabelModel;
	index: number;
}

namespace S {
	export const Label = styled.div`
		display: inline-block;
		position: absolute;
	`;

	export const Foreign = styled.foreignObject`
		pointer-events: none;
		overflow: visible;
	`;
}

export class LabelWidget extends React.Component<LabelWidgetProps> {
	ref: React.RefObject<HTMLDivElement>;

	constructor(props: LabelWidgetProps) {
		super(props);
		this.ref = React.createRef();
	}

	componentDidUpdate() {
		window.requestAnimationFrame(this.calculateLabelPosition);
	}

	componentDidMount() {
		window.requestAnimationFrame(this.calculateLabelPosition);
	}

	findPathAndRelativePositionToRenderLabel = (index: number): { path: SVGPathElement; position: number } | undefined => {
		// an array to hold all path lengths, making sure we hit the DOM only once to fetch this information
		const link = this.props.label.getParent();
		const paths =link?.getRenderedPath();
		if (!link || !paths) {
			throw new Error(`Failed to findPathAndRelativePositionToRenderLabel for index ${index}`);
		}
		const lengths = paths.map((path) => path.getTotalLength()) ?? [];

		// calculate the point where we want to display the label
		let labelPosition =
			lengths?.reduce((previousValue, currentValue) => previousValue + currentValue, 0) *
			(index / (link.getLabels().length + 1));

		// find the path where the label will be rendered and calculate the relative position
		let pathIndex = 0;
		while (pathIndex < paths.length) {
			if (labelPosition - lengths[pathIndex] < 0) {
				return {
					path: paths[pathIndex],
					position: labelPosition
				};
			}

			// keep searching
			labelPosition -= lengths[pathIndex];
			pathIndex++;
		}
	};

	calculateLabelPosition = () => {
		if (!this.ref.current) {
			return;
		}
		const found = this.findPathAndRelativePositionToRenderLabel(this.props.index + 1);
		if (!found) {
			return;
		}

		const { path, position } = found;

		const labelDimensions = {
			width: this.ref.current.offsetWidth ?? 0,
			height: this.ref.current.offsetHeight ?? 0
		};

		const pathCentre = path.getPointAtLength(position);

		const labelCoordinates = {
			x: pathCentre.x - labelDimensions.width / 2 + get(this.props.label.getOptions(), 'offsetX', 0),
			y: pathCentre.y - labelDimensions.height / 2 + get(this.props.label.getOptions(), 'offsetY', 0)
		};

		this.ref.current.style.transform = `translate(${labelCoordinates.x}px, ${labelCoordinates.y}px)`;
	};

	render() {
		const canvas = this.props.engine.getCanvas();

		return (
			<S.Foreign key={this.props.label.getID()} width={canvas?.offsetWidth} height={canvas?.offsetHeight}>
				<S.Label ref={this.ref}>
					{this.props.engine.getFactoryForLabel(this.props.label).generateReactWidget({ model: this.props.label })}
				</S.Label>
			</S.Foreign>
		);
	}
}
