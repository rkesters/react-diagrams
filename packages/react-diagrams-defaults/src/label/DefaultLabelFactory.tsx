import * as React from 'react';
import { DefaultLabelModel } from './DefaultLabelModel';
import { DefaultLabelWidget } from './DefaultLabelWidget';
import { AbstractReactFactory } from '@projectstorm/react-canvas-core';
import { DiagramEngine } from '@projectstorm/react-diagrams-core';

/**
 * @author Dylan Vorster
 */
export class DefaultLabelFactory extends AbstractReactFactory<DefaultLabelModel, DiagramEngine> {
	constructor(engine: DiagramEngine) {
		super('default', engine);
	}

	generateReactWidget(event: {model: DefaultLabelModel}): JSX.Element {
		return <DefaultLabelWidget model={event.model} />;
	}

	generateModel(_event: any): DefaultLabelModel {
		return new DefaultLabelModel();
	}
}
