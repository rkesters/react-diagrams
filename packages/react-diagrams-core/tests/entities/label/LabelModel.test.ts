
import { LabelModel } from '../../../src/entities/label/LabelModel';

describe('LabelModel', () => {

	test.each([
		{
			name: 'default',
			options: { offsetX: 0, offsetY: 0 },
		},
		{
			name: "Undefined Options",
			options: undefined,
			shouldThrow: true
		},
		{
			name: "Missing Offset",
			options: {},
		}
	])('constructor: $name', (tc) => {

		if (tc.shouldThrow) {
			expect(() => new LabelModel(tc.options)).toThrowError();
			return;

		}
		const labelModel = new LabelModel(tc.options);
		expect(labelModel.getOptions()).toMatchSnapshot({ id: expect.any(String) });


	});

	test.each([
		{
			name: 'default',
			options: { offsetX: 0, offsetY: 0 },
			deserial: { extras: {}, id: 'god', locked: true, offsetX: 10, offsetY: 20, selected: true, type: 'test'}
		},
	])('serialize: $name', (tc) => {
		const labelModel = new LabelModel(tc.options);
		const serial = labelModel.serialize();

		expect(serial).toMatchSnapshot({ id: expect.any(String) });

		labelModel.deserialize({ data: {...serial, ...tc.deserial} } as any);

		expect(labelModel.serialize()).toMatchSnapshot();
	});
});
