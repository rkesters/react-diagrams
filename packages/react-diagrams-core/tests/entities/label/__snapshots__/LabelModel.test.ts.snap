// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelModel constructor: Missing Offset 1`] = `
Object {
  "id": Any<String>,
  "offsetX": 0,
  "offsetY": 0,
}
`;

exports[`LabelModel constructor: default 1`] = `
Object {
  "id": Any<String>,
  "offsetX": 0,
  "offsetY": 0,
}
`;

exports[`LabelModel serialize: default 1`] = `
Object {
  "extras": undefined,
  "id": Any<String>,
  "locked": undefined,
  "offsetX": 0,
  "offsetY": 0,
  "selected": undefined,
  "type": undefined,
}
`;

exports[`LabelModel serialize: default 2`] = `
Object {
  "extras": Object {},
  "id": "god",
  "locked": true,
  "offsetX": 10,
  "offsetY": 20,
  "selected": true,
  "type": undefined,
}
`;
